### ISSUE TITLE
GaussianSplatting facade training silently ignores ConfigureOptimizer settings (LR, betas, weight decay) — uses internal updates instead

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1833]
## Problem

Handing a `GaussianSplatting<float>` through the facade —

```csharp
await new AiModelBuilder<float, Tensor<float>, Tensor<float>>()
    .ConfigureDataLoader(loader)
    .ConfigureModel(new GaussianSplatting<float>())
    .ConfigureOptimizer(adam)  // AdamOptimizerOptions { InitialLearningRate = 1e-3, ... }
    .BuildAsync();
```

— appears to train (parameters do move), but the `AdamOptimizerOptions` the caller configured (`InitialLearningRate`, `Beta1`, `Beta2`, `Epsilon`, `WeightDecay`, `UseAdaptiveLearningRate`, ...) are **silently ignored**. Only `MaxIterations` reaches GS. Every other Adam knob is dead code as far as GS is concerned. This violates the "user can fully customize everything" design principle.

Originally filed as "training no-op" — that framing was wrong. Training happens, but via GS's own internal update path, not the facade optimizer.

## Diagnostic data

Reproducer + full instrumentation lives in the discussion thread. Highlights:

| Measurement | Value | Interpretation |
| --- | --- | --- |
| `gs.Layers.Count` | **0** | GS has NO NeuralNetworkBase-tracked layers |
| `gs.ParameterCount` | **0** | Counts `Layers.Sum(ParameterCount)` = 0 |
| `gs.GetParameters().Length` | 472 | Params exist but stored in `_gaussians`, outside standard `Layers` |
| Direct `model.Train` × 10 | 41 ms/step, 16/472 params moved | GS's OWN training path applies updates |
| Facade Adam × 80 steps | 21 ms/step, 16/472 params moved | Same magnitude of movement — facade Adam contributes nothing |
| `adam._lastComputedGradients` after facade run | **length = 0** | `IGradientComputable.ComputeGradients` returns empty vector to facade |

## Root cause

Since GS stores Gaussian params in `_gaussians` (not in `Layers`), the standard facade flow —

1. `AdamOptimizer.Optimize` iterates epochs × batches
2. Per batch: `CalculateGradient(model, x, y)` → the good path (model is `IGradientComputable`) calls `model.ComputeGradients(x, y, LossFunction)`
3. `SynthesizeTapeStepContext` walks `Layers.GetParameterChunks()`, gets an empty context
4. Adam's `Step(context)` runs but has no chunks to update

— never reaches the Gaussian parameters. What DOES update them is GS's own internal splat-gradient code running inside `model.Train` (called from `CalculateGradient` on the same code path). That internal update uses whatever LR / schedule GS's constructor / options were built with; the `AdamOptimizerOptions` from `ConfigureOptimizer` never reach it.

Same architectural pattern affects `DDPMModel` (Week 10 diffusion) — its params come back as a single lumped chunk (`Layers.Count = n/a`, `GetParameterChunks = 1 chunk of 141k params`). Week 10's `DiffusionLab` demo ships with a comment noting the facade doesn't drive DDPM's noise-prediction loop; that's the same customization gap.

## Impact

- **Silent hyperparameter loss.** Every downstream training run through the facade for GS is trained with whatever the GS constructor's defaults are — not what the caller configured.
- **No path to per-parameter-group LR** (paper's positions=1e-4, scales=5e-3, opacity=5e-2, SH=2.5e-3). Reference implementations require this for quality; without customization it's inaccessible.
- **Same pattern lurking for other specialized models** — DDPM confirmed, likely others as AiDotNet expands.

## Fix direction (agreed in design discussion)

**Route `ConfigureOptimizer` settings into the specialized model's internal updater, via an opt-in hook.**

Tiny interface (say `IHyperparameterAware<T>`) that specialized models implement:

```csharp
public interface IHyperparameterAware<T>
{
    // Called by AiModelBuilder immediately before the optimizer's first step.
    // The specialized model reads whatever knobs it needs from the passed
    // optimizer options object and stores them in its internal update state.
    void ApplyOptimizerHyperparameters(IOptimizerOptions<T> options);
}
```

Wiring: extend `AiModelBuilder.ConfigureOptimizer` (or the BuildSupervisedInternalAsync flow it feeds into) so that after the optimizer is finalized but before the first Optimize call, if `model is IHyperparameterAware<T>`, invoke `model.ApplyOptimizerHyperparameters(optimizerOptions)`.

Models opting in (`GaussianSplatting`, `DDPMModel`, future specialized) each implement `ApplyOptimizerHyperparameters` to translate the shared options into their internal schedule (per-attribute LR groups for GS, noise-prediction-specific LR for DDPM, etc.).

## Why this exceeds industry standard

- **PyTorch** and reference GS implementations require users to construct parameter groups manually: `torch.optim.Adam([{params: positions, lr: 1e-4}, ...])`. Users must know GS's per-attribute LR conventions.
- **AiDotNet with this fix**: caller passes one `AdamOptimizerOptions` with base LR; the model's `ApplyOptimizerHyperparameters` derives the per-attribute schedule automatically from the base. Industry-standard defaults live inside the model (users don't need to know GS's magic numbers), but ARE fully customizable (users can override individual attribute LRs via extended options).
- Same pattern eliminates the customization gap for DDPM and any future specialized model — one hook, uniform behavior.

## Wiring surface

`ConfigureOptimizer` only. `AiModelBuilder.ConfigureOptimizer(...)` already takes the options object; extending it to propagate settings to `IHyperparameterAware` models is a purely internal wiring change. No new user-facing Configure method.

## Related

- #1826 / #1829 — original facade unblock chain
- #1832 — different bug (stale ParameterCount cache), same underlying theme (facade + specialized model architectures)

**Comment by ooples** [comment]
Triage note (PR #1789 sweep): reviewed against the #1789 branch. There is related activity on that branch but **no verified fix for what this issue actually asks for**, so it is deliberately left open and not linked for auto-close. Re-check after #1789 merges.

**Comment by ooples** [comment]
### Re-audit against master (`origin/master` @ `a22d7e78bf` (2026-09-05))

**Why this correction exists.** The `Triage note (PR #1789 sweep)` comment on this issue was produced by checking the **#1789 branch**, which was cut on **2026-07-05**. That sweep ran on **2026-08-14**, the day before #1789 merged. Any fix that reached `master` in that 5.5-week window *outside* #1789 was invisible to it, so those verdicts are false-negative-prone. Re-verified below against master itself.

**Verdict: DONE - recommend close**

Shipped by PR #1838 (merged to master as `1ca524d4e2`, 2026-07-10), which the #1789-branch sweep could not see. Master carries explicit provenance:

- `src/Interfaces/IHyperparameterAware.cs:29` - "that specialized models bypass. See #1833 for the full diagnostic."
- `src/AiModelBuilder.BuildPipeline.cs:3144` - "Optimizer options (for #1833 hyperparameter routing + rays-per-batch selection)."
- `src/NeuralRadianceFields/Models/GaussianSplatting.cs:1859` - learning-rate schedule hook, "Called once by AiModelBuilder.BuildAsync"
- `src/Models/Options/GaussianSplattingOptions.cs:72` - documents the `IHyperparameterAware` routing that derives the paper's 1.6e-4 anchor

`ConfigureOptimizer` settings are no longer silently dropped for `GaussianSplatting`.

---
*Posted by an automated re-audit of all 45 open issues against master on 2026-09-05. Every figure above was measured on the tree or read from CI run artifacts; none is carried over from a previous comment.*

**Comment by ooples** [comment]
Closing as fixed and now verified end to end.

**Where the fix lives.** PR #1838 (merged to master as `1ca524d4e2`, 2026-07-10) routed the facade's optimizer options into GS via `IHyperparameterAware`. The 2026-08-14 "no verified fix" verdict was an artifact of sweeping the #1789 branch, which was cut 2026-07-05 and never saw this commit.

**Facade call sites on master `a22d7e78bf`** — both in `AiModelBuilder.BuildPipeline.cs`:
- `:4929` — in the **standard supervised path**, immediately after `finalOptimizer.SetModel(model)` and before any `Optimize` call. This is the path the original repro used.
- `:3149` — in the image-space (#1834) path.

**Existing coverage and its gap.** `GaussianSplattingHyperparameterAwareTests` (5 tests, all passing) calls `ApplyOptimizerHyperparameters` directly. Its own class comment is explicit that it exercises "the interface method directly rather than the full facade round-trip" — so it would keep passing even if the builder stopped invoking the hook entirely. That is precisely the regression #1833 describes.

**Gap now closed.** Added `GaussianSplattingFacadeHyperparameterRoutingTests.BuildAsync_RoutesConfiguredLearningRate_IntoPerAttributeSchedule`, which drives `ConfigureDataLoader` → `ConfigureModel` → `ConfigureOptimizer` → `BuildAsync` and asserts the caller's LR landed.

The fixture is built to discriminate rather than to pass:
- The model is seeded with position `0.05` / color `0.2`, matching neither the library defaults (Adam `1e-3`, base `0.01`) nor the paper anchor (`1.6e-4`).
- The caller passes `InitialLearningRate = 8.0e-4`.
- After the build: position `= 8.0e-4`, color `= 0.0125` (the paper's 15.625 ratio re-derived from the caller's base).
- Explicit `Assert.NotEqual` against the seeded values, so a no-op hook fails the test.

Result: **Passed — 1/1**. Two fixture corrections were needed along the way (GS requires `[N, 6]` ray input and enough rows that the loader's train/val/test split leaves no empty shard); neither indicated a product defect.

The test ships in the consolidated v1 audit PR.
